perfkitbenchmarker/providers/docker/docker_virtual_machine.py

```diff
@@ -0,0 +1,303 @@
+    copy_ssh_command = ['docker', 'cp', self.ssh_public_key,
+                        '%s:/root/.ssh/authorized_keys' % self.name]
+    vm_util.IssueCommand(copy_ssh_command)
+
+    # change ownership of authorized_key file to root in container
+    chown_command = ['docker', 'exec', self.name, 'chown',
+                     'root:root', '/root/.ssh/authorized_keys']
+    vm_util.IssueCommand(chown_command)
+    self._ConfigureProxy()
+
+  def _Delete(self):
+    """Kill and Remove Docker Container"""
+
+    delete_command = ['docker', 'kill', self.name]
+    output = vm_util.IssueCommand(delete_command)
+    logging.info(output[STDOUT].rstrip())
+
+    remove_command = ['docker', 'rm', self.name]
+    output = vm_util.IssueCommand(remove_command)
+    logging.info(output[STDOUT].rstrip())
+
+    return
+
+  @vm_util.Retry(poll_interval=10, max_retries=10)
+  def _Exists(self):
+    """
+    Checks if Container has successful been created an is running
+    """
+
+    info, returnCode = self._GetContainerInfo()
+
+    logging.info("Checking if Docker Container Exists")
+    if len(info) > 0 and returnCode == 0:
+      status = info[0]['State']['Running']
+      if status == "True" or status is True:
+        logging.info("Docker Container %s is up and running.", self.name)
+        return True
+
+    return False
+
+  def _CreateVolumes(self):
+    """
+    Creates volumes for scratch disks. These volumes have to be created
+    BEFORE containers creation because Docker doesn't allow to attach
+    volume to currently running containers.
+    """
+    self.scratch_disks = docker_disk.CreateDisks(self.disk_specs, self.name)
+
+  @vm_util.Retry(poll_interval=10, max_retries=20, log_errors=False)
+  def _DeleteVolumes(self):
+    """
+    Deletes volumes.
+    """
+    for scratch_disk in self.scratch_disks[:]:
+      scratch_disk.Delete()
+      self.scratch_disks.remove(scratch_disk)
+    pass
+
+  def DeleteScratchDisks(self):
+    pass
+
+  def _GetIpAddresses(self):
+    """
+    Sets the internal and external IP address for the Container.
+    """
+    info, returnCode = self._GetContainerInfo()
+    ip = False
+
+    if len(info) > 0 and returnCode == 0:
+      ip = info[0]['NetworkSettings']['IPAddress'].encode('ascii')
+      logging.info("IP: " + str(ip))
+      self.ip_address = ip
+      self.internal_ip = ip
+    else:
+      logging.warning("IP address information not found")
+
+  def _RemoveIfExists(self):
+    if self._Exists():
+      self._Delete()
+
+  def _GetContainerInfo(self):
+    """
+    Gets Container information from Docker Inspect. Returns the information,
+    if there is any and a return code. 0
+    """
+    logging.info("Checking Container Information")
+    inspect_cmd = ['docker', 'inspect', self.name]
+    info, _, returnCode = vm_util.IssueCommand(inspect_cmd, suppress_warning=True)
+    info = json.loads(info)
+    return info, returnCode
+
+  def _ConfigureProxy(self):
+    """
```

Comment:
Single line summary and then elaborate.

https://www.python.org/dev/peps/pep-0257/